src/core/cgroup.c

```diff
@@ -1491,7 +1491,7 @@ static const char *migrate_callback(CGroupMask mask, void *userdata) {
         return unit_get_realized_cgroup_path(userdata, mask);
 }
 
-char *unit_default_cgroup_path(Unit *u) {
+char *unit_default_cgroup_path(const Unit *u) {
```

Comment:
I really don't like this commit. Objects may maintain caches internally, which cannot work if "const" is used, but caches have no effect on whether something is constant or not.

We really should treat "userdata" as entirely opaque, that's kinda the idea of it after all. What the accessors to it do is up to them, and if they use a cache internally, or lazy-load stuff that's entirely fine, and should not prohibited by to many "consts"...

const really is appropriate only for data types we know well, where the data type itself is  API. but that's not really the case for "userdata" stuff, it's supposed to be just some ptr to "something" that is passed around without semantics...

Reply:
Two counterpoints:
1. This is an internal API, and so far none of the functions do any caching like that. If it became necessary to we could always revert this change, it's fairly easy to do with a simple regexp search&replace. No need to avoid const just because it could maybe become a problem in the future.

2. I disagree that we should not use const. The const is saying "this object will not be modified through this pointer". It does not mean that the object will never be modified externally. This would be true even if one of the functions did some caching and changed the internal state of the object. This idea is formally developed in C++, with `mutable` fields and `const_cast<>`, which are about mutating the state through a `const *`. In C++ it's easier because there's a clearer distinction between the object's public API and private API, and the `const*` must not change any state visible through the public API, but changing the internal state is completely OK. In C, there's no formalization of this, since everything is "public", but it allowed to cast away the const. Mutating the object if it is const would be UB, but that's not the case here.